allow back-references in String#sub & #gsub

bjmllr opened the pull request:
Provides #684 

As was suggested, a named argument is used as a flag to enable back-reference scanning. Literal interpretation of the replacement string remains the default.

jhass commented on src/string.cr:
```diff
@@ -1319,11 +1332,77 @@ class String
+  # present in *pattern* or that did not match will be skipped. See `Regex` for
+  # information about capture groups.
+  #
+  # Back-references are expressed in the form `\\d`, where *d* is a group
```
Let's add quotes in the examples, `"\\d"`, so it's less confusing IMO.

jhass commented:
Thinking about this, I'm not sure if the argument is clear on the call side, it's verbose without any information about what it does unless you're familiar with the method. Maybe we can find a different method name? Keep `sub`, `gsub` as is, add `rsub` and `grsub` that does the replacement pattern matching? Admittedly not great names, `r` would be a mnemonic for "replace" or "references".

In any case I'd like to see feature parity for `sub` here :)

bjmllr commented:
Good points. I'll amend with the quotes and new method names.

bjmllr commented:
Restored `#sub` and `#gsub` to their original implementations, added `#rsub` and `#grsub`, placed double-quotes around substrings in docs, and rebased to get CI passing again.

jhass commented:
LGTM, I'd like a second opinion on this though.

ysbaddaden commented:
I like back references so :+1: for the feature! I'm not sure about naming, though. I'd like @asterite to chime in.

asterite commented:
I think we can add this, but I would like just `gsub` and `sub`, which would check back references automatically, unless explicitly told not to do so (so the happy powerful path is turned on by default, but you can make it faster in case it's a bottleneck in your app). The implementation in this PR is OK but can be improved a lot in terms of performance. Right now with this code:

```crystal
time = Time.now
200_000.times do
  "foofoofoofoo".gsub(/o/, "xyzxyzxyzxyzxyzxyzxyzxyzxyzxyzxyzxyz")
end
puts "gsub: #{Time.now - time}"

time = Time.now
200_000.times do
  "foofoofoofoo".grsub(/o/, "xyzxyzxyzxyzxyzxyzxyzxyzxyzxyzxyzxyz")
end
puts "grsub: #{Time.now - time}"
```

I get:

```
gsub: 00:00:00.7161840
grsub: 00:00:01.8086300
```

I was able to reduce it to:

```
gsub: 00:00:00.7063900
grsub: 00:00:00.7627250
```

If the replacement string has a "\\0" in it, the times for grsub go from 1.67s to 1.13s.

For this, grsub first checks if there's a slash in the replacement, otherwise there's no need to check for backreferences over and over. Then I used a `Char::Reader`, because indexing a string is super expensive. Finally, I avoid creating intermediate strings and just append to the only string that's being generated (I added a bunch of code for this, and refactored it).

So, if you'd like so, I can continue working and this and make backreferences work by default, without different overloads.

However... are backreferences that useful? I never used them in my life. And right now you can do them anyway, although in a bit longer way:

```crystal
# proposed
 "foo".rsub(/o/, "x\\0x")

# current way
"foo".sub(/o/) { "x#{$~[0]}x" }

# proposed 
"foo".rsub(/(o)/, "x\\1x")

# current way
"foo".sub(/(o)/) { "x#{$1}x" }
```

So this functionality is already present, you just need to do it in a slightly different way.

ysbaddaden commented:
@asterite here is [my use case](https://github.com/ysbaddaden/frost/blob/master/src/support/inflector.cr#L161). I can't use blocks, because they wouldn't access the `$~` variables, thought I guess I could have `plural(/(hive)\Z/i) { |m| "#{m[1]}s" }` and a macro could abstract it away.

bjmllr commented:
@asterite sorry, you're right, my implementation is not well optimized. Do you want me to keep working on this? It sounds like you've already put significant effort into the implementation, should I update the docs and specs to match the API you have in mind?

I'm not strongly in favor of the feature being accepted, but if it's rejected, then #684 can certainly be closed.

ozra commented:
+1 for back references enabled by default, this is what I would expect. And I favour it much over using "magic globals".

asterite commented:
@bjmllr Oh, no need to say sorry :smile_cat: What I meant is that using your implementation I was able to optimize it a lot very quickly (I reused 95% of your implementation), so the possibility of supporting back references by default sounds like a good idea.

I'll submit a separate pull request that basically does what you did, but slightly faster, and does it with `sub` and `gsub`, with an option to turn it off. I didn't submit any code because I'm still missing the "\\k<var>" case.

@ysbaddaden how does that "\1" work? Shouldn't that be "\\1"?

asterite commented:
@ysbaddaden As a side note, you can assign to `$~` to forward its value: https://play.crystal-lang.org/#/r/tga

It's a bit contrived, but reading from `$~` reads from the variable that was assigned because of a method call inside the current method (basically the variable in scope). But writing to `$~` defines it as a variable **in the caller**. The same applies to `$?`.

The compiler implements this like this: if we assign to `$~` inside a method, the method receives an extra `$~` variable as a pointer argument. If a method calls that method, it declares that variable and passes a pointer to it. You can check that with `--emit llvm-ir`.

bjmllr commented:
Closing in favor of #2253
